src/prpy/planning/vectorfield.py

```diff
@@ -197,32 +219,197 @@ def TerminateMove():
+            twist_perpendicular = util.GeodesicTwist(T_ee_actual,
+                                                     desired_T_ee)
+
+            # The twist tangent to where the end-effector should be
+            # on the goal trajectory
+            # (the feed-forward term)
+            twist_parallel = util.GeodesicTwist(desired_T_ee,
+                                                desired_T_ee_next)
+
+            # Normalize the translational and angular velocity of
+            # the feed-forward twist
+            twist_parallel[0:3] = util.NormalizeVector(twist_parallel[0:3])
+            twist_parallel[3:6] = util.NormalizeVector(twist_parallel[3:6])
+
+            # Apply gains
+            twist = Kp_e*twist_perpendicular + Kp_ff*twist_parallel
+
+            # Calculate joint velocities using an optimized jacobian
+            dqout, _ = util.ComputeJointVelocityFromTwist(robot, twist,
+                                              joint_velocity_limits=numpy.PINF)
+            return dqout
+
+        def TerminateMove():
+            """
+            Function defining the termination condition.
+
+            Fail if deviation larger than position and angular tolerance.
+            Succeed if distance moved is larger than max_distance.
+            Cache and continue if distance moved is larger than distance.
+            """
+            from .exceptions import ConstraintViolationPlanningError
+
+            T_ee_curr = manip.GetEndEffectorTransform()
+
+            # Find where we are on the goal trajectory by finding
+            # the the closest point
+            (_,t,_) = util.GetMinDistanceBetweenTransformAndWorkspaceTraj(
+                                                                     T_ee_curr,
+                                                                     traj)
+            # Get the desired end-effector transform from
+            # the goal trajectory
+            T_ee_desired = openravepy.matrixFromPose(traj.Sample(t)[0:7])
+
+            # Calculate error between current end-effector pose
+            # and where we should be on the goal trajectory
+            geodesic_error = util.GeodesicError(T_ee_desired, T_ee_curr)
+            orientation_error = geodesic_error[3]
+            position_error = geodesic_error[0:3]
+
+            if numpy.fabs(orientation_error) > angular_tolerance:
+                raise ConstraintViolationPlanningError(
+                    'Deviated from orientation constraint.')
+
+            position_deviation = numpy.linalg.norm(position_error)
+            if position_deviation > position_tolerance:
+                raise ConstraintViolationPlanningError(
+                    'Deviated from straight line constraint.')
+
+            # Check if we have reached the end of the goal trajectory
+            dist_to_goal = util.GetGeodesicDistanceBetweenTransforms(T_ee_curr,
+                                                                     T_ee_goal)
```

Comment:
It's strange that constraint deviation is specified separately for position and orientation, but the goal tolerance is specified as a single distance threshold. Both checks should be consistent. I slightly prefer the independent checks, since having separate error messages makes it easier to diagnose why planning failed.